### PR TITLE
[Sketcher] Auto horizontal/vertical

### LINKED ISSUE #10261: [Feature Request] Automatic vertical and horizontal constrain
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Full version info

```shell
[code]
OS: Manjaro Linux (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.21.0.33458 (Git) AppImage
Build type: Release
Branch: master
Hash: 8b89dd29d2382c72d738c52e8974216130ffa9db
Python 3.10.12, Qt 5.15.8, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
Locale: German/Germany (de_DE)
Installed mods: 
  * PieMenu 1.2.7
  * sheetmetal 0.2.63
  * toSketch 1.0.1
  * A2plus 0.4.60n
  * QuickMeasure 2022.10.28
  * Curves 0.6.12
  * CfdOF 1.23.1
  * TabBar
  * IconThemes
  * Glass
  * ProDarkThemePreferencePack 1.0.0
[/code]
```

### Subproject(s) affected?

Sketcher

### Problem description

Hello all,
due to the vertical and horizontal constrain being separate, you have to do a lot of switching back and forth while sketching, which makes the workflow feel chunky. Other CAD programs get around this, by having only one function, which applies the right constrain based on if the line is closer to vertical or horizontal.
In case of FreeCAD, I would leave the existing functions as they are and only add a function on top, which chooses as stated.

Many thanks in Advance

### Anything else?

Here is an example from another CAD programs:
HorizontalVertical_Constraint
horizontal-and-vertical-constraint-in-creo-parametric-sketch

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### LINKED ISSUE #10320: [Feature Request] Group horizontal and vertical constraints
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Full version info

```shell
[code]
OS: Manjaro Linux (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.21.0.33458 (Git) AppImage
Build type: Release
Branch: master
Hash: 8b89dd29d2382c72d738c52e8974216130ffa9db
Python 3.10.12, Qt 5.15.8, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
Locale: German/Germany (de_DE)
Installed mods: 
  * PieMenu 1.2.7
  * sheetmetal 0.2.63
  * toSketch 1.0.1
  * A2plus 0.4.60n
  * QuickMeasure 2022.10.28
  * Curves 0.6.13
  * CfdOF 1.24.4
  * TabBar
  * IconThemes
  * Glass
  * ProDarkThemePreferencePack 1.0.0
[/code]
```

### Subproject(s) affected?

Sketcher

### Problem description

Hello all,
when you work with a rotated view of the sketch, the vertical and horizontal constraints often have switched meanings, which can be disorientating or confusing.
This is just an open discussion for now, but I see two potential ways this can be handled.
First is creating an extra function on top of the current system, which automatically applies a horizontal or vertical constrain based on which is closest (previously seen in #10261).
Second is doing some sort of automatic switching intern the current system, based on the viewing direction. Though I'm not sure how it would actually work and if it would be easier to implement.

Many thanks in advance.

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by PaddleStroke**
Adds a auto horizontal / vertical constraint and command group.
Refactor horizontal/vertical commands which had lot of duplicates.
Fixes #10320 
Fixes #10261

Also participate solving https://github.com/FreeCAD/FreeCAD/issues/11208

### CODE REVIEW

**Comment by maxwxyz** [comment]
nice, is there a new shortcut for the new tool, additionally to "H" and "V"?

**Comment by howie-j** [comment]
Did a quick test, i like it VERY much!

The shortcut is 'A' @maxwxyz 

[Screencast from 2023-11-10 01-15-12.webm](https://github.com/FreeCAD/FreeCAD/assets/82042580/4df653b4-29db-4e91-b9fc-009755d9a9f6)

**Comment by maxwxyz** [comment]
Looks great. What happens when I select multiple lines or horizontal and vertical lines together? Does it also work with multiple points? What happens when you click A on a circle or arc?

**Comment by PaddleStroke** [comment]
Yes it works with multiple lines / points.
For each line / pair of points it will apply the correct constraint.
I put A for shortcut for Alignment. And because it was free. Not sure its the best.

**Comment by maxwxyz** [comment]
A is great!
Does it produce an error if you select geometry that cannot be aligned (arcs, splines,...) or does it just do nothing as today with H/V

**Comment by PaddleStroke** [comment]
Just nothing. It behaves the same
I actually refactored the code of horizontal and vertical so that the 3 tools reuse the same functions. The only difference for the HorVer tool is that it checks orientation and applies the closest.

**Comment by NXTler** [comment]
Oh, I have waited for this for some time now. Thank you very much!

**Comment by maxwxyz** [comment]
@PaddleStroke it seems the icon, especially on the cursor is pretty similar to perpendicular: 
grafik
maybe place the horizontal line on the right side of the vertical with a space between them:  | –
Or maybe a little / abstract grid to indicate its based on the sketch orientation plane?
Or just how the sketch axis look: 
grafik

**Comment by PaddleStroke** [comment]
You are right. Rotating the icon is the simpler. If a design person has a proposal don't hesitate to post. Else I will just rotate 90deg the icon :)

**Comment by pierreporte** [comment]
@PaddleStroke why not a cross like a plus sign? Basically the two constraints fused together.

**Comment by PaddleStroke** [comment]
image
so like this?
It's ok for me

**Comment by pierreporte** [comment]
@PaddleStroke it’s perfect! I like that you can see it’s the two icons together and not just a new one.

**Comment by PaddleStroke** [comment]
@maxwxyz someone asked for a similar tool for parallel / perpendicular. Do you think it is an interesting idea?

**Comment by maxwxyz** [comment]
Definitely would be good! I wonder about the icon of this feature 😄 
Similarly the coincident and point on object could be one tool, depending on the selection of the user I guess.

**Comment by PaddleStroke** [comment]
coincident has already a PR : https://github.com/FreeCAD/FreeCAD/pull/11494
but its going to be added as an option OFF by default for now because it was a bit controversial before. I'll let other battle to get it to be the default.

**Comment by pierreporte** [comment]
> Someone asked for a similar tool for parallel / perpendicular. Do you think it is an interesting idea?

I haven’t seen this in other software. They make this automatically (see #11414). I’m not against but at this point it may be be better to have a button enabling what I propose in #11413.

File: src/Mod/Sketcher/Gui/CommandConstraints.cpp [hunk 1]
```diff
@@ -2725,443 +2721,304 @@ void CmdSketcherConstrainHorizontal::activated(int iMsg)
...
-        static_cast<SketcherGui::ViewProviderSketch*>(getActiveGuiDocument()->getInEdit());
-    Sketcher::SketchObject* Obj = sketchgui->getSketchObject();
-
-    switch (seqIndex) {
-        case 0:// {Edge}
-        {
-            // create the constraint
-            const std::vector<Sketcher::Constraint*>& vals = Obj->Constraints.getValues();
-
-            int CrvId = selSeq.front().GeoId;
```
Comment (PaddleStroke) [inline src/Mod/Sketcher/Gui/CommandConstraints.cpp:3076]: @wwmayer regarding your comment on the other PR, this line of code was already like this before. So the auto vertical/horizontal PR did not changed that.
And it seemed that it wasn't causing problem before. So I'm not sure if it problematic? What do you think?

**Pull request merged**